### PR TITLE
prompt mechanism refactor

### PR DESCRIPTION
**Pull request by laerus**
new PromptFormatter class

renamed `FORMATTER_DICT` to `PROMPT_FIELDS`

some caching when building the prompt (we can do better, some details on #1564)

(some shell refactoring on this matter would prolly come after this)

### CODE REVIEW

**Comment by astronouth7303** [comment]
Notably, this does not change the prompt syntax, as we were discussing elsewhere.

**Comment by scopatz** [comment]
Thanks for putting this in @laerus! Because lot of people use `$FORMATTER_DICT`, I think we need to alias it to the new name and issues a warning about the deprecation when the old name is used for at least one release.  Would you be willing to add that to this? I know it is a pain.... We can do that in a different PR, too, if you want.

**Comment by laerus** [comment]
good point! i'll get to it on this PR

**Comment by laerus** [comment]
this was the easiest way i could find to add the warning, tried with the warnings modules but there is some strange noise.
```
/home/laerus/src/xonsh/xonsh/environ.py:849: UserWarning: FORMATTER_DICT is an alias of PROMPT_FIELDS and will be removed in the next release
  warnings.warn('FORMATTER_DICT is an alias of '
```

and if i change it to `PendingDeprecationWarning` the noise is
```
/home/laerus/.local/bin/xonsh:1: PendingDeprecationWarning: FORMATTER_DICT is an alias of PROMPT_FIELDS and will be removed in the next release
  #!/usr/bin/python3 -u

```

**Comment by laerus** [comment]
i think we can even refactor this to have a function that we can use in case of envvar deprecations in the future.

File: xonsh/prompt/base.py [hunk 1]
```diff
@@ -24,8 +24,70 @@
 from xonsh.prompt.gitstatus import gitstatus_prompt
 
 
+@xt.lazyobject
+def DEFAULT_PROMPT():
+    return default_prompt()
+
+
+class PromptFormatter:
+    """Class used by base_shell"""
```
Comment (scopatz) [inline xonsh/prompt/base.py]: Could this be more descriptive, please?

File: xonsh/prompt/base.py [hunk 2]
```diff
@@ -24,8 +24,70 @@
...
+        self.cache = {}
+
+    def __call__(self, template=DEFAULT_PROMPT, fields=None):
+        """Formats a xonsh prompt template string."""
+        if fields is None:
+            self.fields = builtins.__xonsh_env__.get('PROMPT_FIELDS', PROMPT_FIELDS)
+        else:
+            self.fields = fields
+        try:
+            prompt = self._format_prompt(template=template)
```
Comment (scopatz) [inline xonsh/prompt/base.py]: I think only this line should be in the try-block.  Please move the other two lines to after the except.

**Comment by scopatz** [comment]
Overall, this looks great, @laerus! Just a couple of small cleanups.

File: xonsh/prompt/base.py [hunk 3]
```diff
@@ -24,8 +24,70 @@
...
+            return
+        elif field.startswith('$'):
+            val = builtins.__xonsh_env__[field[1:]]
+            return _format_value(val, spec, conv)
+        elif field in self.fields:
+            val = self._get_field_value(field)
+            return _format_value(val, spec, conv)
+        else:
+            # color or unkown field, return as is
+            return '{{{}}}'.format(field)
```
Comment (laerus) [inline xonsh/prompt/base.py]: i thought string concatenation is slower in python

Comment (laerus) [inline xonsh/prompt/base.py]: concatenation is way faster is this case..

**Pull request merged**